### PR TITLE
feat(client, curriculum): add support for blockLayout property

### PR DESCRIPTION
**Pull request by huyenltnguyen**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

#56090 makes all workshop blocks render in grid layout, but we only want that layout for HTML, CSS, and JS blocks (backend workshops should still render as a list).

This PR adds support for a `blockLayout` property in `meta.json`. We will use this property to determine which layout type a block should be rendered in.

### CODE REVIEW

**Comment by moT01** [comment]
I would be in favor of adding a new property to the meta. Maybe `blockLayout: "grid" | "list" | whatever else we come up with` - one tricky thing is the english challenges, which have a mix of both.

**Comment by ojeytonwilliams** [comment]
> one tricky thing is the english challenges, which have a mix of both.

How about calling it `mixed` for now?

**Review by huyenltnguyen — COMMENTED** [review]
I looked at the code and the English challenges won't be affected by this change.

Their block content is still rendered as a grid (`isGridBased()` returns true):

https://github.com/freeCodeCamp/freeCodeCamp/blob/12026c33f4e1a503193aa7b7060bf6b990ba4315/client/src/utils/curriculum-layout.ts#L6-L11

And the dialogs and tasks rendering are determined one level deeper (`Challenges` component):

https://github.com/freeCodeCamp/freeCodeCamp/blob/12026c33f4e1a503193aa7b7060bf6b990ba4315/client/src/templates/Introduction/components/challenges.tsx#L93-L95

File: curriculum/schema/__snapshots__/challenge-schema.test.js.snap [hunk 1]
```diff
@@ -101,6 +101,16 @@ const schema = Joi.object()
...
+    blockLayout: Joi.when('superBlock', {
+      is: [SuperBlocks.FrontEndDevelopment],
+      then: Joi.valid(
+        'challenge-list',
+        'challenge-grid',
+        'link',
+        'project-list'
+      ).required(),
+      otherwise: Joi.valid(null)
+    }),
```
[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply curriculum/schema/__snapshots__/challenge-schema.test.js.snap:113]: Maybe. Would it be terrible to make `blockLayout` optional and then check for it in the client?

**Review by moT01 — APPROVED** [review]
LGTM 🎉 

Probably want to notify mobile about the schema change.

**Review by ojeytonwilliams — APPROVED** [review]
Hey, Huyen, your changes look good. My only concern is that they build on some  suspect React patterns. 

The main thing is using functions that return JSX instead of functional components. All the newly renamed blocks (ChallengeListBlock et al.) really should be components, imo. It's not that this approach is broken, just that it's misleading. Since they look an awful lot like components, it's far to easy to treat them as such. Then you just add a handy little hook and ooops, broken page.

None of this is because of this PR, though and it would be a big change (and quite a chore), so it's definitely out of scope! I just wanted to mention it for future reference and as a potential refactor when we're feeling strong.

**Comment by ojeytonwilliams** [comment]
> Probably want to notify mobile about the schema change.

Sorry, @Nirajn2311 it's that time again.

**Review by Nirajn2311 — APPROVED** [review]
LGTM from mobile side

**Pull request merged**